pkg/service/endpoints_controller.go

```diff
@@ -17,186 +17,49 @@ limitations under the License.
-			glog.V(5).Infof("endpoints are equal for %s/%s, skipping update", service.Namespace, service.Name)
-			continue
-		}
-		newEndpoints := currentEndpoints
-		newEndpoints.Subsets = subsets
-		newEndpoints.Labels = service.Labels
-
-		if len(currentEndpoints.ResourceVersion) == 0 {
-			// No previous endpoints, create them
-			_, err = e.client.Endpoints(service.Namespace).Create(newEndpoints)
-		} else {
-			// Pre-existing
-			_, err = e.client.Endpoints(service.Namespace).Update(newEndpoints)
-		}
-		if err != nil {
-			glog.Errorf("Error updating endpoints: %v", err)
-			continue
-		}
-	}
-	return resultErr
-}
+var (
+	controllersLock sync.Mutex
+	controllers     = make(map[string]ControllerFunc)
+)
 
-func findDefaultPort(pod *api.Pod, servicePort int, proto api.Protocol) int {
-	for _, container := range pod.Spec.Containers {
-		for _, port := range container.Ports {
-			if port.Protocol == proto {
-				return port.ContainerPort
-			}
-		}
+// RegisterEndpointsController registers a ControllerFunc by name.
+func RegisterEndpointsController(name string, f ControllerFunc) {
+	controllersLock.Lock()
+	defer controllersLock.Unlock()
+	if _, ok := controllers[name]; ok {
+		glog.Fatalf("Endpoints controller %q was registered twice", name)
 	}
-	return servicePort
+	glog.V(1).Infof("Registered endpoints controller %q", name)
+	controllers[name] = f
 }
 
-// findPort locates the container port for the given manifest and portName.
-// If the targetPort is a non-zero number, use that.  If the targetPort is 0 or
-// not specified, use the first defined port with the same protocol.  If no port
-// is defined, use the service's port.  If the targetPort is an empty string use
-// the first defined port with the same protocol.  If no port is defined, use
-// the service's port.  If the targetPort is a non-empty string, look that
-// string up in all named ports in all containers in the target pod.  If no
-// match is found, fail.
-func findPort(pod *api.Pod, svcPort *api.ServicePort) (int, error) {
-	portName := svcPort.TargetPort
-	switch portName.Kind {
-	case util.IntstrString:
-		if len(portName.StrVal) == 0 {
-			return findDefaultPort(pod, svcPort.Port, svcPort.Protocol), nil
-		}
-		name := portName.StrVal
-		for _, container := range pod.Spec.Containers {
-			for _, port := range container.Ports {
-				if port.Name == name && port.Protocol == svcPort.Protocol {
-					return port.ContainerPort, nil
-				}
-			}
-		}
-	case util.IntstrInt:
-		if portName.IntVal == 0 {
-			return findDefaultPort(pod, svcPort.Port, svcPort.Protocol), nil
-		}
-		return portName.IntVal, nil
+// GetEndpointsController creates an EndpointsController using the
+// ControllerFunc registered to the given name. If there is no
+// ControllerFunc registered to the given name, it returns nil.
+func GetEndpointsController(name string, client *client.Client) EndpointsController {
```

Comment:
`func FindEnpointsController(name string, client *client.Client) (EndpointsController, error)` might be more idiomatic with an error of "controller not found with name %s"